driver/src/orderbook/streamed/state.rs

```diff
@@ -155,6 +165,91 @@ impl State {
         // order.
         Ok(())
     }
+
+    fn apply_trade(&mut self, event: &Trade, block_batch_id: BatchId) -> Result<()> {
+        self.apply_trade_internal(
+            event.owner,
+            event.order_id,
+            |order| order.trade(event.executed_sell_amount, block_batch_id),
+            |sell_balance| sell_balance.sell(event.executed_sell_amount, block_batch_id),
+            |buy_balance| buy_balance.buy(event.executed_buy_amount, block_batch_id),
+        )
+    }
+
+    fn apply_trade_reversion(
+        &mut self,
+        event: &TradeReversion,
+        block_batch_id: BatchId,
+    ) -> Result<()> {
+        self.apply_trade_internal(
+            event.owner,
+            event.order_id,
+            |order| order.revert_trade(event.executed_sell_amount, block_batch_id),
+            |sell_balance| sell_balance.revert_sell(event.executed_sell_amount, block_batch_id),
+            |buy_balance| buy_balance.revert_buy(event.executed_buy_amount, block_batch_id),
+        )
+    }
+
+    fn apply_trade_internal(
+        &mut self,
+        user_id: UserId,
+        order_id: OrderId,
+        order_fn: impl FnOnce(&mut Order) -> Result<()>,
+        sell_balance_fn: impl FnOnce(&mut Balance) -> Result<()>,
+        buy_balance_fn: impl FnOnce(&mut Balance) -> Result<()>,
```

Comment:
I am not sure I understand the meaning of these functions as parameters here

Reply:
We want to extract common code from trade and trade_reversion into this function. This is achieved by moving the action that is performed on the order and balances into closures. In `apply_trade` we call `trade` `sell` `buy` and in `apply_trade_reversion` we call `revert_trade` `revert_sell` `revert_buy`.